app/javascript/controllers/dropdate_picker_controller.js

```diff
@@ -0,0 +1,81 @@
+import { Controller } from "@hotwired/stimulus"
+import flatpickr from "flatpickr"
+
+export default class extends Controller {
+  static targets = ["input", "date", "time", "zone", "showing", "editing"]
+
+  connect() {
+    // localize inputs
+    if (this.inputTarget.value) {
+      const date = new Date(this.inputTarget.value)
+      this.setValue(this.dateTarget, date.toLocaleDateString(), true)
+      this.setValue(this.timeTarget, date.toLocaleTimeString(), true)
+    }
+
+    // listen for changes
+    this.dateTarget.dataset.action = `${this.dateTarget.dataset.action} dropdate-picker#change dropdate-picker#defaultTime`
+    this.timeTarget.dataset.action = `${this.timeTarget.dataset.action} dropdate-picker#change`
+    this.zoneTarget.dataset.action = `${this.zoneTarget.dataset.action} dropdate-picker#change`
+
+    // init flatpickr
+    flatpickr(this.dateTarget, {
+      allowInput: true,
+      dateFormat: "n/j/Y",
+    })
+    flatpickr(this.timeTarget, {
+      allowInput: true,
+      dateFormat: "h:i:S K",
+      enableTime: true,
+      enableSeconds: true,
+      noCalendar: true,
+    })
+  }
+
+  editing(event) {
+    event.preventDefault()
+    this.showingTargets.forEach((el) => el.classList.add("d-none"))
+    this.editingTargets.forEach((el) => el.classList.remove("d-none"))
+  }
+
+  change() {
+    if (this.dateTarget.value && isNaN(new Date(this.dateTarget.value))) {
```

Comment:
are these "validations" here because we can't otherwise validate these fields server-side? is the `is-invalid` class mainly for styling to inform the user something is incorrect? does it do anything on submission or otherwise raise an error when the request is made?

Reply:
The server will also throw a rails validation error, if it isn't a valid timestamp.  Or in some cases, rails will clear out a timestamp field if it doesn't recognize the input string?  Or mis-parse it?

The fuzzy nature of how rails parses datetime strings made me want to handle as much as I could upfront.  Fast feedback for the user!